[1단계 - 행운의 로또 미션] 클레어(안은소) 미션 제출합니다.

eunsoA opened the pull request:
안녕하세요 동키콩
로또 미션의 리뷰이 클레어입니다.

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입:

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

1. input에서 validation 이 사용되는 것에 관한 동키콩의 의견이 궁금합니다.
   저는 페어와 논의후, App.js에서 initialize- 라는 내용으로 입력을 받다보니, formatter 형식으로 입력을 받기 위해서는 validation을, UI 로직에서 사용하게 되었습니다. 그러나 validation의 경우, Domain 로직에서 처리하는게 맞다는 생각이 드는데요, 이 경우 validate 를 Domain or UI 어디서 확인하는게 맞을지 동키콩의 생각이 궁금합니다.
- (페어와 이와 관련하여 논의 했던 내용) Domain과 UI 둘다 유효성 검증을 하는게 나을까? 그럼 지금은 ui로직에서만 유효성 검사를 진행하는데, lotto를 생성하는 과정에서도 유효성을 검사를 한번 더 하는게 나을까? 그런데 요구사항에 중복 검사를 하지 말라고 했다. ui에서 validation한 내용을 domain에서 다시 validation하는 것은 중복 검사하는 것을 의미하나?

2. 객체 지향적인 코드인가?
   불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가? 라는 질문에 대해서는 아쉬움이 많이 남습니다. Validation/inputView 이나 app.js 와 같은 부분에서 객체를 잘 사용한 코드다.라는 느낌이 전혀 들지 않는데요. 아직은 익숙치 않고, 배울것이 많아서 그렇다는 생각이 듭니다. 동키콩의 경우, 객체 지향적인 코드를 짜기 위해 어떻게 생각하고 코드를 구현하는지 그 생각의 과정이 궁금합니다.

3. Lotto를 extends한 winningLotto
   winningLotto 를 Lotto에서 extends를 받아서 사용하고 있습니다. Lotto와 winningLotto가 6개의 숫자를 갖는다는 구조가 비슷하기에 그렇게 만들었습니다. 이런 구조를 짠 것에 대해서 동키콩의 의견이 궁금합니다.

4. getNumbers() 사용
   winningLotto 를 Lotto에서 extends를 받아서 사용하고 있습니다. 이 경우 numbers는 private으로 관리하는데, 이 번호를 가져오기 위해서 getNumbers 를 사용했는데요, getter를 사용하면 안된다라는 의견도 있었어서 이에 대해 동키콩의 의견이 궁금합니다.

5. View/Validation/inputView 에 너무 많은 로직이?
     입력 로직에서 validation을 하기 때문에, View/Validation/inputView 로 관리하고 있음. 여기서 너무 많은 validation을 보여주고 있는데 분리되어야 하나? 라는 생각도 드는데 어떤가? 분리되는게 맞을까? 만약 분리된다면 어떻게 분리해야할까?

6. 프로그램 반복의 경우
     retryRun()으로 재귀를 돌게 됨. While()보다 재귀를 도는게 더 나은걸까?

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

1. TDD 적용 방식과 시나리오 순서대로 기능 구현하는 방식 비교
   기존에는 **UI 입력 → 도메인 처리 → UI 출력** 순으로 코드를 작성했지만, 이번에는 **TDD를 적용하여 작은 단위의 도메인부터 구현**하는 방식을 시도했습니다.

   **TDD 적용이 단위 테스트에 유리하다고 생각한 이유**

   - 기능을 작은 단위로 쪼개서 진행하면 **팔로우업(추적)이 용이**합니다.
   - 예상치 못한 문제 발생 시, **원인을 쉽게 찾을 수 있습니다**.
   - 테스트 주도 개발(TDD) 방식과 더 잘 맞는 흐름을 유지할 수 있습니다.
  
  단, **시나리오 순서대로 작성하더라도 세부적으로 분류하면 TDD에 적합한 단위 테스트를 적용할 수도 있다**는 의견도 나누었습니다.

2. eslint 적용
- 이전에는 eslint를 제대로 적용하지 않고 진행헀었는데, 이번에는 eslint의 depth나 var사용에 관한 규칙을 eslint로 적용하였습니다. 이과정에서 버전 문제로 에러가 나서 이를 해결하는데 4-5시간을 썼던 것 같아요.. 🥲

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

1. test와 관련된 커밋 컨벤션의 경우 구조가 수정되면, refactor 로 보내는게 맞는지 궁금합니다. 아니면 test로 보내는게 맞는지 궁금합니다.
   사유 : test와 관련된 건 모두 test 로 보내라고 들었어서 저번 페어 프로그래밍때도 테스트와 관련된 커밋은 모두 test로 보냈는데 동키콩의 의견이 궁금합니다.

2. 하나의 테스트에서 두개의 expect를 날려도 되는지 궁금합니다.


---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계

- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가?
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

이번 리뷰 잘부탁드립니다.
감사합니다 :)

eunsoA committed:
docs: 프로젝트 기능 작업 목록 작성

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 구입 금액에 해당하는 로또 장수를 구하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 1장 가격을 상수로 변경

eunsoA committed:
feat: 로또 1장 발행하는 기능 추가

- 로또 1장은 6개의 번호를 갖음.

Co-authored-by: 신종욱 <dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 장수에 따라 여러 장 발행하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 번호 정렬, 오름차순 정렬 기능 추가

- 로또 번호를 오름차순으로 정렬하는 기능 추가 (Domain)
- 조건에 따라 배열을 받아 번호를 정렬하는 기능 추가 (Utils)

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
refactor: Lotto.test 에서 LottoManager.test로 분리

- 구입 금액에 해당하는 로또 장수를 구한다.
- 로또 장수에 따라 여러 장 발행한다.

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 당첨 번호와 보너스 번호로 당첨 로또를 발행하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 1장당 범위내에서 중복되지 않는 랜덤한 번호 6개를 만드는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 일반 로또 하나와 당첨 로또를 비교해서 번호가 같은 개수를 구하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 일반 로또에 보너스 번호가 있는지 확인하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 당첨 내역을 반환하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 수익률을 반환하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
refactor: Definition 파일명을 definition 으로 수정

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 진입점 추가

- step1-index.js
- App.js

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 구입 금액을 입력받는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 당첨 번호를 입력받는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 보너스 번호를 입력받는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 재시작/종료 여부를 입력받는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 구입 금액 입력에 관한 유효성 검사 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
refactor: utils 내부 readUserInputUntilSuccess 수정

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 당첨 번호 유효성 검사 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 보너스 번호 유효성 검사 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
refactor: initialize를 각각 분리

- initializePurchaseAmount
- initializeWinningNumbers
- initializeBonusNumbers

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 로또 결과 출력하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 구매한 개수만큼 로또 번호 출력 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
feat: 재시작하는 기능 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
docs: 기능 작업 목록 수정

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

eunsoA committed:
docs: 프로젝트 구조 추가

Co-Authored-By: 신종욱 <66900221+dlsxjzld@users.noreply.github.com>

JUDONGHYEOK commented on src/Domain/Model/LottoManager.js:
```diff
@@ -0,0 +1,86 @@
+    return result;
+  }
+
+  #checkCondition(hasBonusNumber, counts) {
```
checkCondition보다 더 좋은 메소드명이 있을 것 같은데 어떻게 생각하세요??

eunsoA replied:
지금 다시보니 checkCondition라는 메소드명은 명확하게 함수의 역할이나 의도가 들어나지 않네요..! 수정했습니다 감사합니다! (commit : [dea0d86](https://github.com/woowacourse/javascript-lotto/pull/370/commits/dea0d863f2e01fcce16d3abbdbdb0f30439be984))

JUDONGHYEOK replied:
오 좋습니다! 추가로 개선을 시도해본다면 개인적으로 이렇게 동일한 구조의 if문이 반복될때는 객체를 활용하는 것도 좋은 방법이 되기도 하는 것 같아요!

eunsoA replied:
오 요 방법은 다시 한번 고민해보겠습니다

JUDONGHYEOK commented on src/Domain/Model/WinningLotto.js:
```diff
@@ -0,0 +1,27 @@
+      lotto.getNumbers().includes(number)
+    ).length;
+    return matchingNumbers;
+  }
```
```suggestion
    countMatchingNumbers(lotto) {
      const winningNumbersSet = new Set(this.getNumbers());
      return lotto.getNumbers().filter(number => winningNumbersSet.has(number)).length;
    }
```
성능과 가독성을 위해 이렇게도 개선할 수 있을 것 같은데 어떻게 생각하세요??

eunsoA replied:
오 코드가 더 간결하고 보기 좋아졌네요!! 

한가지 궁금한 점은 꼭 set()을 해줘야할까요? 저는 Set()을 사용하는 목적은 중복 제거라고 생각했는데요. 
이미 winningNumber의 경우, 유효성 검사를 마친 중복없는 숫자들을 갖고 있기에 꼭 필요한 과정은 아니라고 생각했습니다. 그런데 countMatchingNumbers() 내부에서 Set()을 사용해서 winningNumbersSet을 구하는 한 줄의 이유가 궁금합니다!

```js
countMatchingNumbers(lotto) {
      return lotto.getNumbers().filter(number => this.getNumbers().includes(number)).length;
    }
```

흠 이렇게 쓰고 보니 가독성면에서도 떨어지네요 !

게다가 Array.includes() 와 Set.has()를 비교했을때 시간복잡도 차이가 O(N)과 O(1)로 크네요 ..! 로또 번호가 6개라서 큰 성능 차이는 없겠지만 데이터가 많아지면 Set을 쓰는게 훨씬 낫겠군요

반영해서 수정했습니다! (commit : [b1eb175](https://github.com/woowacourse/javascript-lotto/pull/370/commits/b1eb175e710bf3bc5047db88f564ffe5a1690284))

JUDONGHYEOK replied:
아 사실 winningNumber가 유효성검사를 마친 중복없는 숫자들을 갖고 있다는 사실을 인지하지 못했어요. 이렇게 validation이 도메인이 아닌 view에 위치하게 되면 관리가 어려워지게 되는 것 같아요. 도메인에 종속되는 validation이라면 도메인에 위치시키는 것은 어떨까요? 만약에 또다른 view에서 동일한 Lotto객체를 사용하게 될 때 또다시 validation로직을 중복으로 적어줘야하잖아요? Lotto에 validation책임도 추가되면 어떨지 제안드려봅니다!

JUDONGHYEOK commented on src/Domain/Model/LottoManager.js:
```diff
@@ -0,0 +1,86 @@
+class LottoManager {
+  #lottoList;
+
+  constructor() {}
```
생성자함수를 비워두기 보다는 lottoList를 초기화하는 로직이 있으면 어떨까요?

eunsoA replied:
게임이 반복되니 초기화하는 로직이 있으면 좋겠네요! 요건 생각 못 했네요. 
반영했습니다. 감사합니다! (commit : [e693ea9](https://github.com/woowacourse/javascript-lotto/pull/370/commits/e693ea909f19cd4fc876a64fa1161d01fad88589))

JUDONGHYEOK replied:
좋습니다! 초기화하지 않으면 타입안정성도 떨어지고 makeLottoList를 하기 이전에 접근할 때 타입에러가 발생할 수도 있을 것 같았어요!

JUDONGHYEOK commented on src/Domain/Model/LottoManager.js:
```diff
@@ -0,0 +1,86 @@
+      FOURTH_PRIZE: 0,
+      FIFTH_PRIZE: 0,
+      NONE: 0,
+    };
```
여기서 result를 선언하는 곳에서도 정의하신 상수를 활용하면 어떨까요? 아니면 
```
const result = Object.values(LOTTO_PRIZE_DEFINITION).reduce(
    (acc, prize) => ({ ...acc, [prize]: 0 }),
    {}
  );
```
이런식으로 선언할 수도 있을 것 같아요!

eunsoA replied:
가독성을 생각하면 위의 방식이 낫다는 생각과 시간이 없어서 급하게 위와 같이 코드를 짰었지만 동시에 정의한 LOTTO_PRIZE_DEFINITION를 제대로 못쓴다는 생각이 들었습니다.

그런데 동키콩이 작성해주신걸 보니 보기 더 편하네요 ..! 👍👍
수정해서 반영해두었습니다! (commit : [d173ea3](https://github.com/woowacourse/javascript-lotto/pull/370/commits/d173ea32ae7406f70fbc1eba1836439ead9d759f))

JUDONGHYEOK replied:
오 제가 짠 코드가 괜찮으셨나보군요! 감사합니다!☺️

JUDONGHYEOK commented on src/Domain/Model/LottoManager.js:
```diff
@@ -0,0 +1,86 @@
+      const hasBonusNumber = hasBonusNumbers[index];
+      const lottoResult = this.#checkCondition(hasBonusNumber, counts);
+      result[lottoResult] += 1;
+    });
```
여기 로직들도 조금 단순화 할 수 있지 않을까요? lottoList를 한번 순회할 때 모두 깔끔하게 해결할 수 있지 않을까 ㅅ애각이됩니다!

eunsoA replied:
오 그러네요! 
기존의 코드는 lottoList를 두 번 순회하게 돼서, 이번에는 forEach()를 한 번만 사용하여 countMatchingNumbers()와 checkBonusNumber()를 동시에 실행하도록 수정했습니다.
(commit : [52f9861](https://github.com/woowacourse/javascript-lotto/pull/370/commits/52f9861d3245c06a3fd3e2f331439272dee4bcf0))

JUDONGHYEOK replied:
제눈에는 훨씬 가독성이 올라간 것 같습니다!👍 👍

JUDONGHYEOK commented on __tests__/LottoManager.test.js:
```diff
@@ -0,0 +1,113 @@
+
+test('당첨 내역을 반환한다.', () => {
+  // TODO: 현재는 로또를 로또매니저 내부에서 만들기 때문에 테스트 어려움
+  // -> 로또 머신에서 만들고, 만들어진 로또를 로또 매니저한테 인자로 넘기면 이거 테스트 가능
```
적어주신대로 compareWinnigLotto를 테스트하고자 하셨지만 makeLottoList에서 테스트하기 어려운 random이라는 존재때문에 문제가 되셨을 것 같아요. 이미 문제를 알고 계신 것 같아 다양한 해결책을 비교해보시고 개선해주시길 기대하겠습니다!

JUDONGHYEOK commented on __tests__/LottoManager.test.js:
```diff
@@ -0,0 +1,113 @@
+  expect(prize).toBe(
+    LOTTO_PRIZE_MONEY_DEFINITION.FIRST_PRIZE +
+      LOTTO_PRIZE_MONEY_DEFINITION.THIRD_PRIZE
+  );
```
calculateProfit도 LottoManager의 작업단위이므로 테스트할 필요가 있지 않을까요?

eunsoA replied:
앗 제가 calculateProfit에 관한 테스트는 안만들었었군요.! 
감사합니다! 추가했습니다!
(commit : [fcaaf76](https://github.com/woowacourse/javascript-lotto/pull/370/commits/fcaaf7652d72f58aafe8a769064d7290c07c5243))

JUDONGHYEOK commented on __tests__/LottoManager.test.js:
```diff
@@ -0,0 +1,113 @@
+      min: LOTTO_DEFINITION.MIN_NUMBER,
+      max: LOTTO_DEFINITION.MAX_NUMBER,
+    }
+  );
```
이건 LottoManger에 대한 테스트가 아니라 makeNotDuplicatedRandomNumbers란 유틸함수에 대한 테스트인 것 같아요. 위치에 대한 수정이 필요할 것 같아요!

eunsoA replied:
오 그러네요!
반영해서 수정했습니다 ! 감사해요 :)
(commit : [6439373](https://github.com/woowacourse/javascript-lotto/pull/370/commits/6439373f16fed28e9381106f49f6fe10ac485873))

JUDONGHYEOK commented on __tests__/LottoManager.test.js:
적절하게 유사한 테스트끼리 describe로 묶어줄 수 있을 것 같아요!

eunsoA replied:
앗 1단계 다시보낸 리뷰요청에서 이부분을 빼먹었네요!

JUDONGHYEOK commented on src/View/Constant/query.js:
```diff
@@ -0,0 +1,6 @@
+export const INPUT_MESSAGE = {
+  PURCHASE_AMOUNT: '> 구입금액을 입력해 주세요.',
+  WINNING_NUMBERS: '> 당첨 번호를 입력해 주세요. ',
+  BONUS_NUMBER: '> 보너스 번호를 입력해 주세요. ',
+  RETRY: '> 다시 시작하시겠습니까? (y/n) ',
+};
```
INPUT_MESSAGE를 왜 query로 분리하셨는지 알 수 있을까요??

eunsoA replied:
입출력에서 다뤄지는 상수의 경우, View/Constants에 모았습니다. 
그리고 아래와 같이 분류했습니다.

- errorMessage: 유효성 검사 관련 에러 메시지에서 사용하는 상수
- message: 줄바꿈과 같은 공통으로 사용하는 상수
- query: readLineAsync에서 사용하는 상수

query의 경우, 페어 프로그래밍 중 페어는 입력에 관련된 상수는 query로 항상 빼둔다고 해서 논의끝에 query문으로 분류하였습니다.

JUDONGHYEOK replied:
query라는 파일명이 아무래도 조금 어색한 것 같아요. query라고 하면 데이터 조회를 위해 사용되는 것 같은데 해당 파일내의 정보는 단순히 input창에 보이는 입력 메세지라고 생각이 되어서요.

eunsoA replied:
음 동의해요! 
query라는 단어를 사용한 이유는 페어프로그래밍 당시 페어의 의견에 따랐던 거라 따로 설득하지 않았던 것 같아요. 
2단계에서 수정 반영해보겠습니다!

JUDONGHYEOK commented on src/App.js:
```diff
@@ -0,0 +1,110 @@
+    return retryInput;
+  }
+
+  async run() {
```
run이 너무많은 로직들을 담당하고 있는 것 같아서 적절하게 분리할 수 있을 것 같아요!

eunsoA replied:
네! App.js가 너무 많은 역할을 담당하고 있어서 입력 처리, 검증 로직, 비즈니스 로직으로 분리를 해봤어요.

(1) 검증 로직을 Validator 클래스로 이동 (commit : [a8e1a20](https://github.com/woowacourse/javascript-lotto/pull/370/commits/a8e1a20a1889e13ac09508cd4ab3b2317a6f9ba3))
(2) 사용자 입력 처리 로직을 InputHandler 클래스로 이동 (commit : [74445bc](https://github.com/woowacourse/javascript-lotto/pull/370/commits/74445bc7bcf2e8f0bed305f2b6bed3e8f3d972b6))
(3) 비즈니스 로직을 LottoService 클래스로 이동 (commit : [c192e56](https://github.com/woowacourse/javascript-lotto/pull/370/commits/c192e5611f33531a3959a22f314b1e87435ab473))
(4) App.js는 흐름 제어만 담당하도록 수정

동키콩의 생각은 어떠신가요..?!

JUDONGHYEOK replied:
훨씬 App파일이 간단해졌네요. 이전보다 가독성이 확실히 좋아진 것 같습니다. 😃 그렇다면 다음과 같은 질문을 드릴 수 있을것 같아요!
- validate의 위치는 어디에 위치하는 것이 좋을까요? 현재 모든 validate를 한곳에서 관리하고 InputHandler에서 이를 활용하고 있는데요. validate가 어떤 도메인에 의존적인지 고민해보면 좋을 것 같습니다!
- LottoService를 클래스로 구성하신 이유는 무엇일까요? 이번 미션의 학습목표중 하나가 목적에 맞게 객체와 함수를 활용인만큼 클래스로 구성하신 클레어만의 생각이 궁금합니다!
- 로직은 분리해주셨는데 테스트할 수 있는 부분이 더 늘어난 것 같아요! 테스트를 좀 더 추가해보면 어떨까요??

JUDONGHYEOK commented on src/App.js:
```diff
@@ -0,0 +1,110 @@
+      },
+    });
+    return purchaseAmountInput;
+  }
```
여러 validate로직 때문에 App클래스가 비대해지고 있는 것 같습니다. 객체의 책임을 분리하는 것은 어떻게 생각하세요??

eunsoA replied:
객체의 책임을 분리하고 한가지의 책임을 적절하게 처리하는 것에 대해 너무나도 동의합니다!
다만, 이렇게 App클래스에 하나의 프로그램 전체 로직을 욱여 넣은 이유는 당시 시간이 부족해서 우선은 이렇게 제출했던 것 같아요. 다시 수정해봤습니다!

JUDONGHYEOK commented on src/Utils/math.js:
```diff
@@ -0,0 +1,12 @@
+export const getRandomNumber = (min, max) => {
```
해당 함수는 외부에서 사용하고 있지 않아서 export하지 않아도 될것 같아요!

eunsoA replied:
오 요건 놓쳤던 부분이네요! 감사합니다! (commit : [98f42a0](https://github.com/woowacourse/javascript-lotto/pull/370/commits/98f42a030f85cbf809bf1b95fb5a84632b615e91))

JUDONGHYEOK reviewed (changes requested):
안녕하세요 클레어!! 이번 미션 리뷰어를 맡게된 동키콩입니다. 만나서 반가워요.👏 

도메인과 UI가 적절하게 분리되어 있는 것 같고 신경을 많이 써주신 것 같아요 컨트롤러가 조금 비대하다는 생각이 있지만 충분히 개선할 수 있으시리라 생각됩니다! 또 적절하게 private키워드를 사용하면서 캡슐화를 위해 노력하신 모습도 인상적이였어요!

 개선이 필요한 부분은 LottoManager의 테스트 커버리지가 아쉬운데요. 테스트하기 힘든 구조를 개선하고 커버리지를 올리면 좋을 것 같습니다! 또 App에 대해 너무 많은 책임을 집중되어 있는데요. 적절하게 모델에 분배되면 좋을 것 같아요!

질문에 다답변을 드릴 수 있을지 모르겠지만 하나하나 차근차근 살펴보고 말씀드리면요
## validation관련
제 생각에는 validation의 성격이 중요할 것 같아요 현재는 App에서 모든 validation을 수행하고 있어서 여러 로직들이 혼재되어 있는 것 같은데요. 어떤 validation이냐에 따라 분리할 수 있을 것 같아요. 비즈니스 로직에 가까운 validation은 도메인으로 view에 가까운 validation은 view에서 처리할 수 있을 것 같아요. 예를 들자면 빈문자열 검증이나 사용자 입력형식에 가깝다면 view에서 로또 범위라던지 중복여부와 같은 경우는 도메인에서 처리 가능할 것 같아요!
## 객체지향적 코드
객체지향적인 코드는 무엇이라고 생각하시나요? 저는 보통 객체의 역할과 책임을 많이 생각하는 것 같아요. 이객체가 무엇을 알아야하고 무엇을 해야하는가? 한가지의 책임을 적절하게 처리하고 있는가? 또  더불어 객체간의 협력이 적절하게 이루어지는가 등등 많은 것들이 고려사항이 될 수 있을 것 같아요. 객체지향에 대한 것은 이외에도 수많은 개념들이 존재하기 때문에 차근차근 의식하면서 구현하시면 좋을 것 같습니다!
## 상속
상속은 흔히들 is-a 관계로 표현하잖아요? 위닝로또가 로또의 한 종류로 표현될 수 있기 때문에 적절한 상속이라고 생각해요. 현재는 적절한 구조라고 생각이 됩니다. 다만 상속 외에도 해결할 수 있는 방법들이 있을 것 같기는 해요. 상속의 문제점은 무엇인지 상속보다 합성이라는 말은 왜 나오는지 학습해보면 좋을 것 같습니다!
## getter 사용
흔히들 Tell, Don't Ask 원칙이라고 하죠. 직접적인 상태 보다는 필요한 정보만을 알려주는 형태가 객체지향적으로 좋은 코드라고 여겨지는데요. 사실 그렇게 하면 좋지만 지켜지기 어려운 것도 사실인 것 같아요. 이정도의 getter는 괜찮지 않나 개인적으로 생각해요. 지금 getter의 문제점은 객체상태를 그대로 반환하고 있어서 사실상 외부에서 데이터 변경이 가능해요. 때문에 복사를 해서 데이터조작이 불가하도록 하는 캡슐화정도는 필요하다고 생각합니다.
## 재귀
지금 구조에서 큰차이가 있다고 생각하진 않지만 일반적으로 제가 느끼기로는 while문이 특정 알고리즘을 제외하고는 가독성이 좋다고 생각하고 크진않지만 메모리 효율도 조금 더 좋아서 while문을 선호하기는 합니다

## 커밋 컨벤션
개인적으로는 구조 수정이 테스트와 관련되면 test:로, 로직 개선이 주 목적이면 refactor:로 커밋하는 것이 적절하다고 생각해요. 하지만 적절한 팀 컨벤션이 있다면 어느쪽이든 괜찮다고 생각합니다.
## 두개의 expect
한 테스트에서 여러 expect를 사용하는 것은 가능합니다. 다만, 하나의 테스트는 하나의 동작이나 결과를 검증하는 것이 좋죠. 때문에 논리적으로 연관된 검증이라면 여러 expect를 사용해도 괜찮다고 생각해요.


클레어 로또 미션도 너무너무 수고많으셨습니다! 지금 코드도 좋지만 LottoManager에 대한 테스트 코드는 꼭 추가되었으면 해요! 기간이 많이 남아있고 개선점들도 보여서 Request changes하도록 하겠습니다~~

eunsoA committed:
refactor: checkCondition 메서드명을 countLottoResult로 변경

- 리뷰어의 피드백 반영
(https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965599819)

eunsoA committed:
refactor: LottoManger에서 calculateLotto로 분리

- LottoManger에서 lottoList와 관련없는 계산 로직은 calculateLotto로 분리함.
- calculateLotto 내부 메서드명을 calculateLotto-로 통일함.

eunsoA committed:
refactor: countMatchingNumbers 성능 개선

- Set 활용하여 성능 개선

- 리뷰어 피드백 반영
(https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965608038)

eunsoA committed:
refactor: compareWinningLotto의 result 초기화 방식 개선

- 리뷰 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965621199)

eunsoA committed:
refactor: lottoList 순회 로직 수정

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965624618)

eunsoA committed:
refactor: 불필요한 export 제거

- getRandomNumber 함수가 외부에서 사용되지 않으므로 export 키워드 제거

- 리뷰어 피드백 반영
(https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965683831)

eunsoA committed:
refactor: lottoManager 생성시 초기화 로직 추가

- 리뷰어 피드백 반영
(https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965610098)

eunsoA committed:
test: LottoManger에서 calculateLotto로 분리에 따른 테스트 에러 수정

eunsoA committed:
test: LottoManager.test.js 에서 calculateLotto.test.js로 분리

eunsoA committed:
refactor: app에서 유효성 검증 로직을 Validator로 분리

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965681147)

eunsoA committed:
refactor: 사용자 입력 처리 로직을 InputHandler 클래스로 이동

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965681147)

eunsoA committed:
refactor: View 하위 utils 로직을 Utils 폴더로 이동

- readLineAsync.js
- utils.js

eunsoA committed:
feat: 비즈니스 로직을 LottoService 클래스로 이동

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965681147)

eunsoA committed:
feat: getter에서 사용하는 얕은 복사를 깊은 복사로 수정

- Lotto.js

eunsoA committed:
refactor: LottoManger.test 에서 math.test로 분리

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965665454)

eunsoA committed:
fix: 리팩터링 중 오류 해결

- 구조 변경에 따른 import문 수정
- 매개변수 2개 이하 조건만족을 위해 LottoService 함수 분리
- getBonusNumber 매개변수 누락 수정

eunsoA committed:
test: 수익률 반환에 대한 테스트 추가

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965663313)

eunsoA committed:
refactor: getter에서 사용하는 얕은 복사를 깊은 복사로 수정

eunsoA committed:
refactor: LottoManager의 랜덤 생성에 따른 테스트 추가

- 리뷰어 피드백 반영 (https://github.com/woowacourse/javascript-lotto/pull/370#discussion_r1965660403)

eunsoA committed:
docs: 프로젝트 구조 수정

eunsoA commented:
안녕하세요 동키콩!
우선 1단계 리뷰를 아주 자세히 열심히 작성해주셔서 감사합니다!

1단계 리뷰요청 후 수정하면서 제가 신경썼던 점은 아래와 같습니다.

1. 현재 getter에서 객체 상태를 그대로 반환하는 내용을 수정.
- getter의 얕은 복사를 깊은 복사로 변경하였습니다.

2. 상속
- 이번 미션에서는 Lotto와 WinningLotto의 상속의 구조를 유지하면서 2단계를 진행하고, 상속으로 인해 문제점이 생기는지 알아보고 싶어요.  + 여유가 된다면 합성이랑 비교해볼게요.

3. App 클래스가 비대해지는 문제 해결
- 현재 App.js가 너무 많은 역할을 담당하고 있어서 입력 처리, 검증 로직, 비즈니스 로직으로 분리를 해봤어요. 
- (1) 검증 로직을 Validator 클래스로 이동 
- (2) 사용자 입력 처리 로직을 InputHandler 클래스로 이동 
- (3) 비즈니스 로직을 LottoService 클래스로 이동 
- App.js는 흐름 제어만 담당하도록 수정

4. LottoManager에 대한 테스트 코드 추가. 
- 랜던번호 생성하는 함수를 넘겨받도록 코드를 수정하여 테스트가 가능하도록 리팩터링하였는데요, 이런 코드에 대한 동키콩의 생각이 궁금해요. 어떤 단점이 있을지, 궁금합니다!

추가로 혹시 리뷰 과정에서 제가 커밋링크를 잘못달았다거나, 건강한 리뷰 문화를 위해 알아야하는 점이 있다며 거리낌없이 말씀해주세요:)

그럼, 좋은 하루 보내세요!
감사합니다! 🙇🙇

JUDONGHYEOK commented:
안녕하세요! 클레어!! 대부분의 리뷰를 잘 반영해주셔서 감사드립니다!!🙇🏻‍♂️ 확실히 구조적으로 더 깔끔해지고 가독성이나 유지보수성의 측면에서 더 나은 코드가 된 것 같네요! 구조적인 면에서 조금더 개선할 수 있는 사항은 코멘트로 달아놓았으니 확인부탁드려요! 또 더 개선할 수 있을 것 같은 부분은 테스트인데요. 현재 테스트 케이스를 추가해볼 수 있을 만한 부분들이 조금씩 눈에 보이는 것 같아요! 이부분에 대해서 조금 더 고민해보시고 추가해주시면 더 안정감 있는 애플리케이션이 되지 않을까 생각합니다!!

1. getter에서의 깊은복사
- 아주 좋은 것 같습니다! 객체를 그대로 반환하는 것은 사실상 원본을 그대로 노출시켜 데이터를 변경할 수 있게 하는 것과 동일한데요. 진정한 캡슐화가 진행된 것 같네요!
2. 상속
- 직접 느껴보는 것 만큼 좋은 학습이 없다고 생각해요! 2단계에서 상속의 장점과 단점에 대해서 클레어 만의 생각을 꼭 정리해보길 바래요!
3. App 클래스 문제
- 요건 코멘트로 달아놓았습니다!
4. 로또 매니저에 대한 테스트 코드 추가
- 랜덤로직이 적절하게 분리된 것 같아요! 복잡도가 조금 늘어나긴 했어도 이번미션의 주제가 테스트이기 때문에 테스트코드를 작성하는데 수월해지고 생성함수에 대한 전략을 분리함으로써 유연성이 확보되었다고 생각해요! 앞으로도 적절한 트레이드오프를 클레어만의 생각으로 분리해서 적용하면 좋을 것 같습니다.

2단계에서 꼭 테스트 코드를 추가해주시리라 믿고 어프로브&머지하도록 하겠습니다! 다양한 시도해주셔서 재밌게 리뷰했어요 감사하고 수고하셨습니다 클레어!!🤓
